mythril_core/src/acpi/madt.rs

```diff
@@ -0,0 +1,335 @@
+use super::rsdt::SDT;
+use crate::error::{Error, Result};
+use bitflags::bitflags;
+use byteorder::{ByteOrder, NativeEndian};
+use core::convert::TryFrom;
+use core::fmt;
+use core::ops::Range;
+
+/// See Table 5-43 in the ACPI spcification.
+///
+/// Note that these offsets are relative to the end of the
+/// SDT (the end of the Creator Revision at offset 36).
+mod offsets {
+    use super::*;
+    /// 32-bit physical address at which each processor can access its
+    /// local APIC.
+    pub const LOCAL_INT_CTRL_ADDR: Range<usize> = 0..4;
+    /// Multiple APIC Flags.
+    pub const FLAGS: Range<usize> = 4..8;
+    /// Interrupt Controller Structures
+    pub const INT_CTRL_STRUCTS: usize = 8;
+}
+
+/// Interrupt Controller Structure Type Values
+///
+/// See Table 5-45 in the APIC specification.
+#[repr(u8)]
+#[derive(Clone, Copy, Debug)]
+enum IcsType {
+    ProcessorLocalApic = 0x00,
+    IoApic = 0x01,
+    InterruptSourceOverride = 0x02,
+    NmiSource = 0x03,
+    LocalApicNmi = 0x04,
+    LocalApicAddressOverride = 0x05,
+    PlatformInterruptSource = 0x08,
+    ProcessorLocalX2Apic = 0x09,
+    LocalX2ApicNmi = 0x0a,
+}
+
+impl IcsType {
+    /// Expected length of buffer for ICS type.
+    ///
+    /// See the Structure definition tables found in
+    /// `ACPI § 5.2.12` for details.
+    pub fn expected_len(&self) -> usize {
+        match *self {
+            IcsType::ProcessorLocalApic => 8,
+            IcsType::IoApic => 12,
+            IcsType::InterruptSourceOverride => 10,
+            IcsType::NmiSource => 8,
+            IcsType::LocalApicNmi => 6,
+            IcsType::LocalApicAddressOverride => 12,
+            IcsType::PlatformInterruptSource => 16,
+            IcsType::ProcessorLocalX2Apic => 16,
+            IcsType::LocalX2ApicNmi => 12,
+        }
+    }
+
+    /// Check the length of bytes available for the given
+    /// ICS type.
+    pub fn check_len(&self, length: usize) -> Result<()> {
+        // The length includes the type and length bytes.
+        if length == self.expected_len() - 2 {
+            Ok(())
+        } else {
+            Err(Error::InvalidValue(format!(
+                "Invalid length={} for type=0x{:x}",
+                *self as u8, length
+            )))
+        }
+    }
+}
+
+impl TryFrom<u8> for IcsType {
```

Comment:
I think we can use `TryFromPrimitive` for this, like we did here (among other places): https://github.com/mythril-hypervisor/mythril/blob/e843004df4b5ec8d5ea0c0f20d64037d1ac2fead/mythril_core/src/error.rs#L8

Reply:
I'll give it a shot. Good suggestion.